### PR TITLE
Freshbooks integration

### PR DESCRIPTION
**Pull request by ra-ver**
- Created new invoicing service: invoice-api. This interacts with Freshbooks to create, view and email Invoices
- Added new api end points in billing-api that act as interface for invoice-api end points for all the clients
- Other code changes in billing-api to make it to enable support for invoicing

### CODE REVIEW

File: invoice-api/src/service/freshbooks.ts [hunk 1]
```diff
@@ -0,0 +1,503 @@
...
+        for(var idx in emails){
+            if(adminEmails.indexOf(emails[idx].toLowerCase()) <0){
+                output.push(emails[idx]);
+            }
+        }
+        return output;
+
+    }
+
+    // assumes date is : null or in 'YYYY-MM-DD' format
```
Comment (andricDu) [inline invoice-api/src/service/freshbooks.ts]: block comments on top of functions

File: invoice-api/src/service/freshbooks.ts [hunk 2]
```diff
@@ -0,0 +1,503 @@
...
+        this.agent = new https.Agent({
+            rejectUnauthorized: config.rejectInsecure
+        });
+        this.authenticator = authenticator;
+    }
+
+
+    /*
+    Steps to send Invoice:
+        1. Get bearer token using the refresh token
```
Comment (andricDu) [inline invoice-api/src/service/freshbooks.ts]: 👍  great comment!

File: invoice-api/src/service/freshbooks.ts [hunk 3]
```diff
@@ -0,0 +1,503 @@
...
+            pageIdx++;
+        }// while loop ends here
+       return allCustomersData;
+    }
+
+    private async getAllCustomersPaged(pageNumber: number, email:string): Promise<any> {
+        return axios.get(`${ this.apiConfig.api }/accounting/account/${ this.apiConfig.account_id }/users/clients?include[]=recent_contacts&search[email_like]=${ email }&page=${ pageNumber }&per_page=100`,
+            {headers: this.headers, httpsAgent: this.agent })
+            .then( response => {
+                console.log(response.data);
```
Comment (andricDu) [inline invoice-api/src/service/freshbooks.ts]: you might want to remove the logging. Could be pretty verbose.

File: invoice-api/src/service/freshbooks.ts [hunk 4]
```diff
@@ -0,0 +1,503 @@
...
+            });
+        });
+        return flattenedInovicesJson;
+    };
+
+    private async getAllCustomers(userEmail:string): Promise<any>{
+        userEmail = (userEmail == null || userEmail == "") ?  "":userEmail;
+        let pageCount = 1; let pageIdx = 1;
+        let allCustomersData = [];
+        while(pageCount  >= pageIdx){
```
Comment (andricDu) [inline invoice-api/src/service/freshbooks.ts]: at first glance the while loop takes a second to understand. Another issue is that it has to make each request sequentially, waiting for the next one to come back before moving forward. 

A better way would be to just fire off all the requests you'll need in parallel, and maintain an array of promises which you can just `await` on with `Promise.all()`

Something like this (my syntax might be wrong):

```javascript
    private async getAllCustomersParallel(userEmail:string): Promise<any>{
        userEmail = (userEmail == null || userEmail == "") ?  "" : userEmail;
        let promises = [];
        let pagedCustomerResults = await this.getAllCustomersPaged(1, userEmail);
        
        for (let i=1; i<pagedCustomerResults.pages; i++) {
            promises.push(this.getAllCustomersPaged(i, userEmail));
        }
        
        let responses = await Promise.all(promises);
        return responses.reduce((a,b) => a.concat(b));
    }
```

File: invoice-api/src/service/freshbooks.ts [hunk 5]
```diff
@@ -0,0 +1,503 @@
...
+            }
+        } // contacts iteration ends here
+        return false;
+    }
+
+    private async findInvoice(invoiceNumber:string): Promise<any> {
+        return axios.get(`${ this.apiConfig.api }/accounting/account/${ this.apiConfig.account_id }/invoices/invoices?search[invoice_number]=${invoiceNumber}`,
+            {headers: this.headers, httpsAgent: this.agent })
+            .then( response => {
+                console.log(response.data);
```
Comment (andricDu) [inline invoice-api/src/service/freshbooks.ts]: might be too verbose to log.

File: invoice-api/src/service/freshbooks.ts [hunk 6]
```diff
@@ -0,0 +1,503 @@
...
+            }).catch(err => {
+                throw new Error(err.response.statusText);
+            })
+    }
+
+    private async getCustomerInfo(customerId:string){
+        return axios.get(`${ this.apiConfig.api }/accounting/account/${ this.apiConfig.account_id }/users/clients?search[userid]=${customerId}&include[]=recent_contacts`,
+            {headers: this.headers, httpsAgent: this.agent })
+            .then( response => {
+                console.log(response.data);
```
Comment (andricDu) [inline invoice-api/src/service/freshbooks.ts]: another one here, will probably end up being too verbose. Also without timestamps reducing the usefulness.

File: invoice-api/src/service/freshbooks.ts [hunk 7]
```diff
@@ -0,0 +1,503 @@
...
+            // form search string for customer ids
+            for(let idx in customerIDs){
+                searchStr += '&search[customerids][]='+customerIDs[idx]
+            }
+        }
+        // Freshbooks limits the max number of responses to 100 per page; hence requesting 100
+        return axios.get(`${ this.apiConfig.api }/accounting/account/${ this.apiConfig.account_id }/invoices/invoices?${ searchStr }&page=${ pageNumber }&include[]=lines&per_page=100`,
+            {headers: this.headers, httpsAgent: this.agent })
+            .then( response => {
+                console.log(response.data);
```
Comment (andricDu) [inline invoice-api/src/service/freshbooks.ts]: another log entry we can probably do away with.

File: invoice-api/src/server.ts [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+        }
+        let fbService = new FreshbooksService(config['freshbooksConfig'], req.app.get('settings').authenticator);
+        fbService.emailExistingInvoice(email,invoiceNumber).then({} = () => {
+            res.send("Invoice emailed.");
+        }).catch(err => {
+            res.status(500).send(err);
+        });
+    });
+
+
```
Comment (andricDu) [inline invoice-api/src/server.ts]: you can probably get rid of one or two of these newlines.

File: billing-invoice/src/service/invoice.ts [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+        });
+
+    };
+
+    private async getInvoicesSummaryData(){
+        let currentDate = new Date();
+        let dateText = currentDate.toISOString().slice(0,10);
+        return axios.get(`${ this.apiRootURL }/getAllInvoices?date=${ dateText }`)
+            .then( response => {
+                //console.log(response.data);
```
Comment (andricDu) [inline billing-invoice/src/service/invoice.ts]: delete this.

File: invoice-api/src/service/freshbooks.ts (hunk 4 above)
Comment (cheapsteak) [inline invoice-api/src/service/freshbooks.ts]: give this a shot

```js
private async getAllCustomersParallel(userEmail:string): Promise<any>{
    let customers = await this.getAllCustomersPaged(1, userEmail || "");
    if (customers.pages > 1) {
      customerResults = customerResults.concat(
        await Promise.all(_.range(1, customers.pages).map(i =>
          this.getAllCustomersPaged(i, userEmail || "")))
      )
    };
    return customers;
}
```

**Comment by cheapsteak** [comment]
If we want to keep the ability to log around, but only display it when required, would recommend giving this library a look

https://www.npmjs.com/package/debug

Currently the most download package on npm
http://npmcharts.com/compare/lodash,request,async,chalk,express,bluebird,commander,underscore,debug

File: invoice-api/src/service/freshbooks.ts [hunk 10]
```diff
@@ -0,0 +1,503 @@
...
+                'invoice_number' : item.invoice_number,
+                'payment_status' : item.payment_status,
+                'invoice_status' : item.v3_status,
+                'cpu_cost' : this.getLineItemValue("CPU",item.lines),
+                'image_cost' : this.getLineItemValue("Image",item.lines),
+                'volume_cost' : this.getLineItemValue("Volume",item.lines),
+                'discount' : item.discount_value,
+                'total' : item.amount.amount
+            });
+        });
```
Comment (cheapsteak) [inline invoice-api/src/service/freshbooks.ts]: Small nitpick, would favor maps over initialize, iterate, and push

```js
let flattenedInovicesJson = allInvoicesForThisPeriod.map(item => (item) => ({
      'current_organization' : item.current_organization,
      'date' : item.create_date,
      'invoice_number' : item.invoice_number,
      'payment_status' : item.payment_status,
      'invoice_status' : item.v3_status,
      'cpu_cost' : this.getLineItemValue("CPU",item.lines),
      'image_cost' : this.getLineItemValue("Image",item.lines),
      'volume_cost' : this.getLineItemValue("Volume",item.lines),
      'discount' : item.discount_value,
      'total' : item.amount.amount
  })
);
```

File: invoice-api/src/service/freshbooks.ts [hunk 11]
```diff
@@ -0,0 +1,503 @@
...
+
+    }
+
+    // assumes date is : null or in 'YYYY-MM-DD' format
+    public async getInvoicesSummaryData(date:string, user:any, admin:boolean) {
+        await this.authenticate();
+        let minDate = null;
+        let maxDate = null;
+        minDate = date;
+        maxDate = new Date();
```
Comment (cheapsteak) [inline invoice-api/src/service/freshbooks.ts]:
```diff
- let minDate = null;
- let maxDate = null;
- minDate = date;
- maxDate = new Date();
+ let minDate = date;
+ let maxDate = new Date();
```

File: invoice-api/src/service/freshbooks.ts [hunk 12]
```diff
@@ -0,0 +1,503 @@
...
+            throw Error("error: User account: "+ customerEmail +" does not have access to this Invoice: "+invoiceNumber );
+        let emailInfoJson = this.getExistingInvoiceEmailInfo(customerEmail,invoiceNumber);
+        return this.emailInvoice(invoiceInfo.id,emailInfoJson);
+    }
+
+    private validEmailForCustomer(email:string, customerInfo:any): boolean{
+        let lowerCasedEmail = email.toLowerCase();
+        if(customerInfo.email != ''){
+            if(customerInfo.email.toLowerCase() == lowerCasedEmail) return true;
+        }
```
Comment (cheapsteak) [inline invoice-api/src/service/freshbooks.ts]: Can this be

```js
if (customerInfo.email && customerInfo.email.toLowerCase() == lowerCasedEmail) {
 return true;
}
```
?

File: invoice-api/src/service/freshbooks.ts [hunk 13]
```diff
@@ -0,0 +1,503 @@
...
+        // seach for email in secondary contacts
+        let contact:any;
+        for(let contactIdx in customerInfo.recent_contacts){
+            contact = customerInfo.recent_contacts[contactIdx];
+            if(contact.email != ''){
+                if(contact.email.toLowerCase() == lowerCasedEmail) return true;
+            }
+        } // contacts iteration ends here
+        return false;
+    }
```
Comment (cheapsteak) [inline invoice-api/src/service/freshbooks.ts]: Can this be

```js
return !!_.find(
  customerInfo.recent_contacts,
  contact => contact.email && contact.email.toLowerCase() == lowerCasedEmail
);
```
?

Comment (cheapsteak) [inline invoice-api/src/service/freshbooks.ts]: Actually I think the entire function can be reduced to this:

```js
private validEmailForCustomer(email:string, customerInfo:any): boolean{
  return _.every([
    email !== '',
    customerInfo.email.toLowerCase() == lowerCasedEmail,
    !!_.find(
      customerInfo.recent_contacts,
      contact => contact.email && contact.email.toLowerCase() == lowerCasedEmail
    )
  ]);
}
```

**Comment by ra-ver** [comment]
Addressed all PR comments and added features/changes requested by Finance and others. Some of Sid's changes are merged in this branch as we were testing it earlier on my machine. However; his latest changes are not integrated in this PR. He raised a separate PR for those changes.

**Pull request merged**